### PR TITLE
RFC add foreach to iterators

### PR DESCRIPTION
**Pull request by larroy**

### CODE REVIEW

**Comment by tshepang** [comment]
[rendered view](https://github.com/larroy/rfcs/blob/master/text/0000-foreach.md)

File: text/0000-foreach.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+# Motivation
+
+After a chain of transformations on an iterator it might be desirable to have a side effecting
+operation that consumes the iterator. Adding .foreach to iterators is proposed, which is a common
+idiom in other languages for performing this operation. While the same effect can be achieved with a
+for loop, in general is considered that flat is better than nested and leads to more readable code.
+
+# Detailed design
+
+The design is simple, just syntacting suggar around a loop that consumes the iterator as seen in
```
Comment (erickt) [inline text/0000-foreach.md]: I believe you mean "syntactic sugar" here.

**Comment by sinistersnare** [comment]
i think `foreach` does not work well with rust naming conventions, so I would say either `for_each`, or just `each`. I like `each`.

**Comment by blaenk** [comment]
I too prefer `each`, as in Ruby.

**Comment by jsanders** [comment]
I think this would be really nice. :+1: to `each` from me too.

Side note: I had to do some clicking around to figure out what is being proposed here. Could you put the main idea in the RFC itself?

**Comment by mneumann** [comment]
+1 ```each```

**Comment by SimonSapin** [comment]
This is called `apply` in rust-itertools: http://bluss.github.io/rust-itertools/doc/itertools/trait.Itertools.html#method.apply

**Comment by SimonSapin** [comment]
However, since `for` loops are expressions, it’s not clear to me how a method doing the same is better.

**Comment by jsanders** [comment]
@SimonSapin: I had the same thought at first, but [this comment on reddit](http://www.reddit.com/r/rust/comments/2s5jjs/does_having_foreach_implemented_for_iterators/cnmhloc) convinced me that it can be quite nice to end a chain of iterator methods with an application of a closure instead of splitting it up into a separate iterator chain plus a for loop. But maybe it really just belongs in itertools, where it already is.

**Comment by SimonSapin** [comment]
While this is interesting, I think the way to go at this point is to demonstrate wide usage in a Cargo crate like itertools. The bar should be high to move thing into std.

**Comment by blaenk** [comment]
I don't think that would be the best metric for measuring desire in this feature. I personally think this would be _very_ convenient, but not to the point where I would necessarily bring in a separate dependency just for that.

I agree that the bar should be high, but this is a very simple addition, and one that is a core part of many other languages that have similar iteration/chaining/streams APIs (such as [Scala](http://www.scala-lang.org/api/current/index.html#scala.collection.parallel.ParIterableLike$Foreach), [C#](http://msdn.microsoft.com/en-us/library/bwabdf9z(v=vs.110).aspx), [Java 8](http://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#forEach-java.util.function.Consumer-), and [Ruby](http://www.ruby-doc.org/core-2.2.0/Array.html#method-i-each) to name a few).

I think the RFC would benefit from including a code example motivating the addition of this method. While I think for loops are great, I do think it's nice to not have to break the chain of iteration methods just to force the side-effecting changes to go through.

``` rust
let mut somechain = xs.iter().take(5).filter(|x| x.is_thing()).map(|x| x + 1);

for thing in somechain {
  // do side effecting operation
  println!("{:?}", x);
}
```

It would be nicer to be able to just do this for certain short cases:

``` rust
xs.iter()
  .take(5)
  .filter(|x| x.is_thing())
  .map(|x| x + 1)
  .each(|x| println!("{:?}", x));
```

I'm not adamant about including this, but it did always strike me as odd that this method was missing from iterators.

**Comment by SimonSapin** [comment]
> I personally think this would be very convenient, but not to the point where I would necessarily bring in a separate dependency just for that.

I used to think that way. And then I learned to love Cargo. Publishing and using very small crates like [`matches!`](https://github.com/SimonSapin/rust-std-candidates/blob/master/matches/lib.rs) is fine! If you’re already using Cargo, what is *really* the cost of an additional dependency?

**Comment by SimonSapin** [comment]
That said some things still *should* go into std eventually. But in the mean time, real world usage experience can bring feedback of changes you might want to do before setting thing is backward-compatibility stone.

**Comment by blaenk** [comment]
@SimonSapin Oh that's not what I meant. I don't mind bringing in another dependency if the thing I want to use is necessary and useful. What I meant was that in my case, and I bet other people, if this method were available to me out of the box I would _readily_ use it, but otherwise I don't care enough about it to warrant bringing in a dependency simply for it.

Take your `matches!` macro for example, I loved it and was disappointed when it wasn't accepted in the std. There are times when I would have liked to have used it, but to date I haven't bothered with actually bringing in that dependency, simply because I don't care enough, and so far it hasn't seemed overly inconvenient to just write it out explicitly. But, again, if it _were_ there then I _would_ use it.

With that in mind, it's also why I wouldn't mind too much if this weren't accepted, but it does feel like it's conspicuously missing from the list of iterator methods to others who are familiar with similar iterator APIs.

And I still agree with you that we should be careful/cautious about the things that are proposed to be added to the std; I wholeheartedly agree. I think it's a good idea to, like you say, first get feedback about such proposals, but I think that's already what this RFC is for; it's not really an earth-shattering proposal with deep and complex ramifications, AFAIK.

We do have to draw the line somewhere. After all, if we accept this, then what's not to stop us from adding a bunch of other simple proposals until the std is undirect and bloated, right? Sure, but my argument is that this method is a very common component of similar APIs, I think of it as an iteration 'primitive' (such as `map`, `fold`, etc.) which has precedent in other established languages.

@larroy: Please include a code example showing specifically how this may be useful. Feel free to use [mine](https://github.com/rust-lang/rfcs/pull/582#issuecomment-70168089) or a better one you could come up with. It's nice to have self-contained RFCs so people could easily get a full picture of what you intend, even if the use of your proposed method seems very obvious to you.

**Comment by thestinger** [comment]
This already existed (it was called `advance`) and after a lot of deliberation if was removed because it's almost entirely redundant with `for` loops. Rust has tried to be a clean language with one way to do things. It has to contend with the need to expose lots of flexibility for performance reasons, but I think it has done very well so far - and part of that is because it has rejected redundant methods like this.

**Comment by thestinger** [comment]
@blaenk: Your code example is just deliberately making the `for` loop look more verbose. Nothing stops you from avoiding a temporary in both cases while preserving a clean coding style.

**Comment by blaenk** [comment]
@strcat: It's not some sleight of hand attempt to deceive anyone to show that the temporary is necessary; everyone knows the temporary isn't necessary. I realized I should have changed or clarified it for this very reason but I had to go AFK before I got a chance.

The reason I separated it out in the example was because that's what I normally do when the chain gets very long, though in that example it's not particularly long. When the chain is very long I like to have each segment on its own line for readability, and as a result it makes the for loop look very weird, so I separate it out that way.

**Comment by thestinger** [comment]
It looks perfectly fine without a temporary when following the "official" Rust coding style.

**Comment by mneumann** [comment]
@thestinger: when it fits in one line, it's ok:

```rust
for x in xs.iter().take(5).filter(|x| x.is_thing()).map(|x| x + 1) {
  // do side effecting operation
  println!("{:?}", x);
}
```

But if you want to break it up, line by line, it's getting pretty ugly IMHO and rightward:

```rust
for x in xs.iter()
           .take(5)
           .filter(|x| x.is_thing())
           .map(|x| x + 1) {
  // do side effecting operation
  println!("{:?}", x);
}
```

From the visual point of view, and that's my subjective opinion, having an ```each``` method whatever it's name might be, be it ```foreach```, ```exec``` or ```do```, is much more attractive. People coming from Ruby will just miss it, because there you have both ```for``` and ```each``` and both are useful on it's own (e.g. ```each``` often being used when it fits in one line, while ```for``` for multi-line code).

**Comment by Ericson2314** [comment]
Does `iter.map(|x| ...).skip_while(|_| true)` achieve the same thing?

**Comment by P1start** [comment]
Even shorter is `iter.map(|x| ...).count()`, which is the same as `foreach` but returns the number of iterations as well (which can easily be ignored).

**Comment by nagisa** [comment]
@Ericson2314 @P1start I mentioned on the discuss thread, that 
```rust
iterator.fold((), |_, x|{ 
    // for_each body;
})
```
is equivalent to proposed function and is much clearer in intent compared to `.map(…).anything()`.

**Comment by huonw** [comment]
FWIW, there's also `all` (which is exactly the same as the old `advance`) and `any`.

**Comment by pnkfelix** [comment]
So, it seems like debate has mostly settled down at this point.  (From what I can tell, we have two camps: One saying that a method like this is worthwhile since its name signals "my callback carries a side-effecting operation", and the other saying "you should just be writing this as a for-loop, not pure method composition."  Well, I guess there's also a third camp, perhaps a subcamp of the second, that says "if you really want to do this without making a for-loop, you can use pre-existing methods for it"; but that is not really a counter-argument since the first camp sees intrinsic value in the name `for_each`.)

Regarding @SimonSapin 's point about Cargo: I believe the core team agrees with him, and wants to encourage things that can go into third party libraries to actually go up on Cargo, regardless of whether they are tiny/trivial things.  (The main reason not to put this up on Cargo is if we wanted to encourage people to use it, but I suspect that in general we want to encourage people to use for-loops instead, all other things being equal.)

----

The RFC seems to have been abandoned by its author, since they have not responded to requests from the community for the addition of a motivating example to the RFC itself.

@larroy am I mistaken about this?

**Comment by larroy** [comment]
Something similar has been added to itertools http://bluss.github.io/rust-itertools/doc/itertools/trait.Itertools.html

**Comment by pnkfelix** [comment]
Ah, great! and `itertools` is up on `cargo`.

Here's the method itself: http://bluss.github.io/rust-itertools/doc/itertools/trait.Itertools.html#method.foreach

@larroy you said "similar"; are the semantics of the above method different than what you proposed here, to your knowledge?

In any case, I'm inclined at this point to close this ticket.

**Comment by larroy** [comment]
Well, if there's no intention to put it in rust then we can close as it's already in itertools for those who care.

**Comment by aturon** [comment]
As per the most recent discussion, I'm going to go ahead and close this RFC.

**Pull request closed without merging**